Update ONNX docs

lewtun opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR gives the ONNX part of the documentation a much needed update to:

* Explain why we have three ONNX configuration objects (`OnnxConfig`, `OnnxConfigWithPast`, and `OnnxSeq2SeqConfigWithPast`) and how to interpret them.
* Explain what we mean by the `--features` argument of the export CLI.
* Provide an end-to-end guide on how to export a custom model, using the DistilBERT implementation as an example. I chose this approach so readers can run actual code instead of relying on a hypothetical unsupported architecture.

I also removed the deprecated section that was based on the `convert_graph_to_onnx.py` script as it seems to confuse people on which API to use. I couldn't find clear instructions on how the deprecation cycle works for the documentation of scripts like `convert_graph_to_onnx.py`, so please let me know if I should re-instate this part of the docs.

I also tweaked the help messages in the `transformers.onnx` CLI to make things a bit clearer to the end user.

The diff on the docs is quite huge, so I recommend reading the file directly.

lewtun committed:
Remove docs for deprecated ONNX export

lewtun committed:
Tidy up the CLI help messages

lewtun committed:
Revamp ONNX docs

lewtun committed:
Update auto-config table

lewtun committed:
Use DistilBERT as example for consistency

lewtun committed:
Wrap up first pass at ONNX docs

lewtun committed:
Fix table check

lewtun committed:
Add tweaks and introduction

lewtun commented on docs/source/serialization.mdx:
```diff
@@ -10,27 +10,38 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-unified and community-driven format to store and, by extension, efficiently execute neural network leveraging a variety
-of hardware and dedicated optimizations.
+Once exported, a model can optimized for inference via techniques such as
+quantization and pruning. If you are interested in optimizing your models to run
```
I point people to `optimum` here to indicate that `transformers` is only concerned with _exporting_ models, not optimising them.

lewtun commented on docs/source/serialization.mdx:
```diff
@@ -49,255 +60,365 @@ Ready-made configurations include the following models:
-                ("token_type_ids", {0: "batch", 1: "sequence"}),
-            ]
-        )
+| Feature                              | Auto Class                           |
```
We might want to automatically populate this table based on the state of `FeaturesManager`. I've left it as manual for now.

lewtun committed:
Add cross-ref

lewtun commented on docs/source/serialization.mdx:
```diff
@@ -49,255 +60,365 @@ Ready-made configurations include the following models:
 
-- **Change the target opset version of the generated model.** (More recent opset generally supports more operators and
-  enables faster inference)
+* Encoder-based models inherit from [`OnnxConfig`]
```
I'm not entirely sure how cross-referencing works with the new doc - looking at other sections, it seems square brackets are used. Since `OnnxConfig` is not visible in the docs page, perhaps I should not do this here?

lewtun replied:
I've opted to link to the source code for these classes as they're not currently visible on the docs. (Will add that in a separate PR)

lewtun committed:
Fix missing import

lewtun commented on docs/source/serialization.mdx:
```diff
@@ -49,255 +60,365 @@ Ready-made configurations include the following models:
 
-This conversion is handled with the PyTorch version of models - it, therefore, requires PyTorch to be installed. If you
-would like to be able to convert from TensorFlow, please let us know by opening an issue.
+The ONNX conversion is supported for the PyTorch versions of the models. If you
```
I'm not 100% sure whether we don't support TensorFlow or not since we technically do in the deprecated `convert_graph_to_onnx.py` script

michaelbenayoun replied:
I think that it is not well defined for now, we need to work on the export from TF models, but for now I think it is best to keep this.

lewtun committed:
Merge branch 'master' into update-onnx-docs

lewtun committed:
Fix style

michaelbenayoun commented on docs/source/serialization.mdx:
```diff
@@ -49,257 +60,365 @@ Ready-made configurations include the following models:
-On the other hand, after the training phase, it has been shown one can greatly reduce the range and the precision of
-*float32* numbers without changing the performances of the neural network.
+Notice that the outputs property follows the same structure as the inputs; it
+returns an `OrderedDict` of named outputs and their shapes. The output structure
```
It is actually not always an `OrderedDict` I think, it depends on the configuration (as we did not always make sure to use that), but the important takeaway is that the inputs and outputs must be fed in the proper order to the export function.

lewtun replied:
Thanks for catching this! I've now replaced `OrderedDict` with "mapping" in this sentence and added a tip box below to clarify the role of the `OrderedDict` in this particular example.

michaelbenayoun reviewed (approved):
Concise and clear, that's really great, thanks @lewtun !!

LysandreJik commented:
This is great, thanks for working on it @lewtun! Pinging @stevhliu and @sgugger for knowledge.

lewtun committed:
Add permalinks to ONNX configs

lewtun committed:
Clarify role of OrderedDict

lewtun committed:
Merge branch 'master' into update-onnx-docs

sgugger commented on docs/source/serialization.mdx:
````diff
@@ -50,257 +61,374 @@ Ready-made configurations include the following models:
-configuration of each model. For example, for BERT:
+which outputs:
 
 ```python
````
There shouldn't be `python` markers for outputs in the doc.

sgugger commented on docs/source/serialization.mdx:
````diff
@@ -50,257 +61,374 @@ Ready-made configurations include the following models:
+ONNX operator set that will be used during the export:
+
+```python
+# Returns 11
+print(onnx_config.default_onnx_opset)
````
```suggestion
>>> print(onnx_config.default_onnx_opset)
11
```
The doctest syntax (used everywhere else in the doc ;-) ) makes it easier to show the output.

sgugger commented on docs/source/serialization.mdx:
````diff
@@ -50,257 +61,374 @@ Ready-made configurations include the following models:
-format. By default, weights of a neural network are stored as single-precision float (*float32*) which can express a
-wide-range of floating-point numbers with decent precision. These properties are especially interesting at training
-where you want fine-grained representation.
+which gives:
+
+```python
+OrderedDict([("last_hidden_state", {0: "batch", 1: "sequence"})])
+```
````
Sme as above, use the doctest syntax instead.

sgugger commented on docs/source/serialization.mdx:
```diff
@@ -50,257 +61,374 @@ Ready-made configurations include the following models:
 
-The quantization process will infer the parameter *scale* and *zero_point* from the neural network parameters
+All of the base properties and methods associated with
+[`OnnxConfig`](https://github.com/huggingface/transformers/blob/c4fa908fa98c3d538462c537d29b7613dd71306e/src/transformers/onnx/config.py#L52)
```
```suggestion
[`OnnxConfig`]
```
Let the documentation resolve the link to this class maybe? Or did you really need the source code here?

lewtun replied:
Yeah, I added the permalink because I wasn't sure if the doc will resolve the link. But you're right, I'll remove the URL for now :)

sgugger replied:
It's actually not in the API documentation so you should either add it or keep the link to GitHub :-/

lewtun replied:
Hehe, OK I'll re-instate the link for now and add the API documentation in a separate PR to keep this one focused on just the user guide.

sgugger commented on utils/check_table.py:
```diff
@@ -207,11 +207,11 @@ def get_onnx_model_list():
         filename=os.path.join(PATH_TO_DOCS, "serialization.mdx"),
         start_prompt="<!--This table is automatically generated by make style, do not fill manually!-->",
-        end_prompt="This conversion is handled with the PyTorch version of models ",
+        end_prompt="The ONNX conversion is supported for the PyTorch versions of the models.",
```
Thanks for updating this!

sgugger reviewed (approved):
Great, thanks a lot for making this doc better!

lewtun committed:
Update docs/source/serialization.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

lewtun committed:
Add doctest syntax to code blocks

lewtun committed:
Remove permalinks

lewtun committed:
Revert "Remove permalinks"

This reverts commit 099701daf0db27823457867938efdb2d4f22a7c1.

lewtun commented:
Thank you for the feedback @sgugger 🙏 !

I've now added the doctest `>>>` syntax to all the Python code blocks and removed the permalinks to the class references. OK to merge once the CI tests pass?
